### PR TITLE
Animation issue in Setting Modal

### LINKED ISSUE #1358: Inconsistencies Between Animation Speeds Amongst Various Platforms
**Platform - version:**  All (Cross platform) - version 1.0.1-389

**Action Performed (reproducible steps):**
1. Click the Profile Image icon at the top of the sidebar on the left.
1. A settings modal should animate in on the right.
1. Close the settings modal
1. Observe the differences in animation time across platforms.

**Expected Result:** The animation should behave the same across all platforms.

**Actual Result:**
On iOS, the transition is smooth (300ms in and out as expected):
105801273-a62cb380-5f4d-11eb-98e2-3146fddbd890

However, on Web, the out transition is virtually indetectable:
Kapture 2021-01-28 at 10 18 11

Inconsistencies of a similar nature exist between iOS/Android as well.

**Notes/Photos/Videos:** Added above
 
**Logs - JS/Android/iOS (if applicable)**: N/A

### PR DESCRIPTION
**Pull request by parasharrajat**
@yuwenmemon @shawnborton  @marcaaron Please review.
### Details
 The transition effect on the Setting modal is not the same on different devices and TransitionOut is not detectable.

### Fixed Issues
Fixes #1358 

### Tests
Run the app on a mobile device and observe the animation while opening and closing the settings modal. Now open the application on the web browser and observe the animation effect. It should look the same.

#### Steps:
1. Click the avatar on the left sidebar.
2. Observe the opening animation.
3. Close the setting screen by either clicking the close icon or clicking backdrop.
4. Observe the closing animation.
5. Now repeat 1-4 for different platforms and compare the animation.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop

#### Web
ezgif com-video-to-gif

#### Mobile Web
ezgif com-video-to-gif(1)

#### Desktop
ezgif com-video-to-gif(2)

### CODE REVIEW

**Comment by parasharrajat** [comment]
Hey Guys, 
I have checked and found one issue. While we resize the app lower than the mobile resolution on the web the app breaks. The cause of this issue is that the animation is not being built and attached dynamically. After thorough testing, I found that `react-native-modal` supports dynamic animations but its equivalent `modal-enhanced-react-native-web` which we are using for the web platform lacks that code. I have successfully added that into the `modal-enhanced-react-native-web` package (Check the PR Dekoruma/react-native-web-modal#43 ) but I think the maintainer of that repo is not responsive. Thus is it possible we can use our own copy of this package where we can use my modified version?

**Comment by marcaaron** [comment]
>I have successfully added that into the modal-enhanced-react-native-web package (Check the PR Dekoruma/react-native-web-modal#43 ) but I think the maintainer of that repo is not responsive. Thus is it possible we can use our own copy of this package where we can use my modified version?

Ah hmm that's interesting. I didn't realize we were using this package on web instead of `react-native-modal`.

I'm not sure we'll want to fork this repository, but if it's the only option then we'd need to fork it first to our organization and then merge your changes in there.

Perhaps we can wait a bit and see if they respond to that PR or we can reach out directly and nudge them. It has only been a day so maybe they will still respond.

**Comment by parasharrajat** [comment]
@marcaaron  Check here. I was thinking the same initially. But then I come to know this after hours of fooling around with `react-native-modal` code.

https://github.com/Expensify/Expensify.cash/blob/d4ba77f3732b422b0f6b26740da489af28982d21/config/webpack/webpack.common.js#L113-L120

**Comment by marcaaron** [comment]
Thanks for the context. Yes, I think the best thing to do is wait for a bit and try to get your code merged into `modal-enhanced-react-native-web` if possible.

**Comment by parasharrajat** [comment]
I agree. But the last update was in 2018. We better mail the guy. :smile:

**Comment by marcaaron** [comment]
Looking into this a bit more and I think there is a possibility we are just doing something wrong. 

Can we try to fix this with styling somehow? If the maintainer doesn't respond then we've decided to only fork the repository as a last resort (and most likely we won't do it). I did some testing with this library on another branch and think it should work correctly under normal circumstances.

**Comment by parasharrajat** [comment]
So this lib is very old fork of react-native-modal for the web and almost 90% code is same but react-native-modal is updated more often but this fork has not been updated since 2018. I have picked the necessary code from react-native-modal to patch PR changes but I am sure a lot can be changed.

You can just pick both and compare to see what part of code is still behind from the react-native-modal.

**Comment by marcaaron** [comment]
@parasharrajat, I understand what you are trying to do here. But I am suggesting we explore a better solution without forking the library if you cannot get these changes merged. We are not interested in maintaining a fork of this library.

**Comment by parasharrajat** [comment]
Will explore more for some solution

**Comment by parasharrajat** [comment]
Perfect, I removed the `modal-enhanced-react-native-web package`. There are two reasons.

1. We just don't need a separate package for the web as `react-native-modal` compiles to the web.
2. `modal-enhanced-react-native-web` is outdated and not maintained. The maintainer is suggesting to use other [here](https://github.com/Dekoruma/react-native-web-modal#announcement).

Now react-native-modal works for both web and native.

Please review.

**Review by yuwenmemon — CHANGES_REQUESTED** [review]
Thanks, @parasharrajat looks good and tests well! Can you also please test this on android and iOS? I tested on Android and it seems we still have issues with the animation timing on that platform. Thank you!

**Comment by marcaaron** [comment]
Just chiming in to say that I think the animation issue's root cause may be related to how many chats you have in your list and probably indicate a performance issue somewhere. 

**Observe:**

`master` with only one chat
https://user-images.githubusercontent.com/32969087/106976730-07e2df80-66fd-11eb-93b2-617baca34c0e.mp4

`master` with many chats
https://user-images.githubusercontent.com/32969087/106976742-0c0efd00-66fd-11eb-8c83-f3785178ccc6.mp4

`parasharrajat:parasharrajat-animation` with many chats
https://user-images.githubusercontent.com/32969087/106976776-17fabf00-66fd-11eb-9d82-292000e24def.mp4

**Comment by parasharrajat** [comment]
I will dig this out why its behaving that way.

**Comment by parasharrajat** [comment]
So, Surely it's a performance issue, If you use `useNativeDriver` then we get smooth animations. 
## Before
https://user-images.githubusercontent.com/24370807/106992084-a4f55500-679d-11eb-8d60-a936b04e092a.mp4

## After
https://user-images.githubusercontent.com/24370807/106992075-9dce4700-679d-11eb-8e66-2da68edd21f1.mp4


Now one thing worth mentioning is that the Swipe gesture is not behaving the same. Look at the before the video, swipe happens back and forth, and follows your finger. But in after the video, the swipe happens in one go, and it does not follow your finger.

FPS of the js thread is pretty low when we run the animations or open the Setting the modal. (it drops to 4 or 6 from 59.3 on opening the setting modal). This is the main issue, I think as per the article [here](https://reactnative.dev/docs/performance#dropping-js-thread-fps-because-of-doing-a-lot-of-work-on-the-javascript-thread-at-the-same-time). Animations are calculated on the JS thread so I am sure it can't be smooth at 4 fps.

### Tip
1. First thing, I noticed is that `withWindowDimensions` wrapper component is bad for performance as it updates while animations are running, on the other hand, `useWindowDimensions` does not update which boost the fps to 12 from 4 on my device.

**Comment by marcaaron** [comment]
Sorry, I'm not really sure what `useNativeDriver` has to do with anything. The issue states the animation inconsistency is with the _web_ implementation. In `react-native-web` there is only one thread for animations to run on.

>I noticed is that withWindowDimensions wrapper component is bad for performance as it updates while animations are running

How did you verify this?

**Comment by parasharrajat** [comment]
@marcaaron  Web issue is resolved it works fine now. But animation is not at all showing on my android device as you mentioned https://github.com/Expensify/Expensify.cash/pull/1367#issuecomment-773716385. 
`useNativeDriver` is basically helping on android. 

> I noticed is that withWindowDimensions wrapper component is bad for performance as it updates while animations are running

I used `useEffect` hooks to log the props that are coming from the `withWindowDimensions`, so when we run the animation you will see a series of logs with the props value which means props are changing.
```
useEffect(() => {
console.debug(props.windowWidth);
}, [props.windowWidth]);

```

On the other hand, when I used `useWindowDimensions`, there was no output when the animation was running.
I may be wrong :smile:

**Comment by marcaaron** [comment]
Ok thanks for the explanation. I think presence of a re-render itself does not necessarily indicate a performance issue. On the other hand re-rendering something that is expensive to re-render might. So I think we are re-rendering something (probably `OptionRow` items) and that is causing some performance issue. But it could be any number of things.

**Comment by marcaaron** [comment]
>I am not sure if this comes under the scope of this issue

If in fixing web we break Android then I'd say it maybe falls under the scope 😄 

Anyway, I tested your suggestion to try `useNativeDriver` on Android and it works well. But note that it seems to have a bug on iOS where there is a strange flash. See docs here: https://github.com/react-native-modal/react-native-modal#the-modal-flashes-in-a-weird-way-when-animating

Other than that, I like the solution as presented. The performance issues I experienced are caused by unnecessary `OptionRow` re-renders and can be fixed by modifying the `memo()` `areEqual` like so...

https://github.com/Expensify/Expensify.cash/blob/ed64a9d59b84e7d0107287ff4e5fbefc40dd1b81/src/pages/home/sidebar/OptionRow.js#L160-L195

Maybe there is something else to look into but so far this seems to resolve things for me even with many many chats in the list so perhaps we can incorporate the above changes for now.

**Comment by parasharrajat** [comment]
@marcaaron  @yuwenmemon 
I have patched the following things for now

1. Added `useNativeDriver` for android only.
2. Removed the if condition here as it destroys the node on the route change thus no exit animations for the new Group Page.
https://github.com/Expensify/Expensify.cash/pull/1367/files#diff-1407d7e65cda4bcbcccbdfc4ee27d5950ea06bf3a94a153ce3eccf1a08c222e9

File: src/styles/getModalStyles/index.android.js
Comment (marcaaron) [inline src/styles/getModalStyles/index.android.js:1]: Hmm I think we are duplicating a lot of logic here by copying this entire file. Better to just create a structure like this

```
.
└── components/
    └── Modal/
        ├── index.js - imports Modal
        ├── index.android.js - adds useNativeDriver
        └── Modal.js
```

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
@marcaaron Let me know if this is what you want. Thanks

File: src/components/Modal/ModalSkeleton.js
Comment (marcaaron) [inline src/components/Modal/ModalSkeleton.js:46]: I would maybe remove this. We can make it so all Android modal `useNativeDriver` and this would keep things a lot simpler.

File: src/components/Modal/index.android.js
Comment (marcaaron) [inline src/components/Modal/index.android.js:58]: This can just be called `Modal` no need to call it a `Skeleton`

File: src/styles/getModalStyles.js
Comment (marcaaron) [inline src/styles/getModalStyles.js:145]: I think we should move this setting out of this file entirely.

File: src/components/Modal/index.js
Comment (marcaaron) [inline src/components/Modal/index.js:8]: These propTypes can go into a new file and get imported in each of the `index.js` and `index.android.js` files

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by marcaaron** [comment]
@parasharrajat left a couple more comments and you have conflicts to fix. overall solution looks good to me.

File: src/components/Modal/ModalSkeleton.js
Comment (parasharrajat) [reply src/components/Modal/ModalSkeleton.js:46]: Ok, So there are multiple types of modal. IF we make a true in the `index.android.js` then it will be true for all the types of the modal. Do you want that? If so, We can follow your next comment and remove the `useNativeDriver` from the `getModalStyles.js`.

File: src/components/Modal/index.android.js
Comment (parasharrajat) [reply src/components/Modal/index.android.js:58]: Then, there would be two components with the same name

File: src/components/Modal/ModalSkeleton.js
Comment (marcaaron) [reply src/components/Modal/ModalSkeleton.js:46]: Yes

File: src/components/Modal/index.android.js
Comment (marcaaron) [reply src/components/Modal/index.android.js:58]: maybe name it `BaseModal` similar to what we do here?
Screen Shot 2021-02-06 at 9 36 21 AM

**Comment by parasharrajat** [comment]
Updated

**Review by marcaaron — APPROVED** [review]
LGTM. The performance issue will be handled in a separate PR.

**Comment by marcaaron** [comment]
@stitesExpensify @yuwenmemon feel free to give this a review/test.

**Review by yuwenmemon — APPROVED** [review]
Thanks @parasharrajat

**Pull request merged**
